### PR TITLE
Reuse retrieve.Cache for customEnvParser in secret.go

### PR DESCRIPTION
**Pull request by Avni-Sharma**
Fix for https://github.com/redhat-developer/service-binding-operator/issues/272

### CODE REVIEW

File: pkg/controller/servicebindingrequest/retriever_test.go [hunk 1]
```diff
@@ -207,7 +204,7 @@ func TestCustomEnvParser(t *testing.T) {
 	crName := "db-testing"
 
 	f := mocks.NewFake(t, ns)
-	f.AddMockedUnstructuredCSV("csv")
+	// f.AddMockedUnstructuredCSV("csv")
```
Comment (isutton) [inline pkg/controller/servicebindingrequest/retriever_test.go]: Remove comment.

**Comment by sbose78** [comment]
Avni, could you test this feature end to end and provide the detailed steps here on the PR ?

File: pkg/controller/servicebindingrequest/reconciler.go [hunk 2]
```diff
@@ -352,7 +353,7 @@ func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 
 	logger.Debug("Retrieving data to create intermediate secret.")
 	retriever := NewRetriever(r.dynClient, plan, sbr.Spec.EnvVarPrefix)
-	retrievedData, err := retriever.Retrieve()
+	retrievedData, retrievedCache, err := retriever.Retrieve()
```
Comment (otaviof) [inline pkg/controller/servicebindingrequest/reconciler.go]: In the current design of the operator, `Retriever` should only fetch data specified by the `Plan` (from `Planner`), therefore, `Retreive` should only return one single set of data, not two. The idea is the later components won't have to figure out sources of data anymore, and act in a single view of the binding context data.

File: pkg/controller/servicebindingrequest/reconciler.go [hunk 3]
```diff
@@ -374,5 +375,5 @@ func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 	}
 
 	logger.Info("Starting the bind of application(s) with backing service...")
-	return r.bind(logger, binder, secret, retrievedData, sbr, sbrStatus, objectsToAnnotate)
+	return r.bind(logger, binder, secret, retrievedData, retrievedCache, sbr, sbrStatus, objectsToAnnotate)
```
Comment (otaviof) [inline pkg/controller/servicebindingrequest/reconciler.go]: Same idea here, make sure `retreievedData` contains it all.

File: pkg/controller/servicebindingrequest/retriever.go [hunk 4]
```diff
@@ -267,19 +267,19 @@ func (r *Retriever) store(key string, value []byte) {
...
 // "bindable resources", gathering extra data. It can return error on retrieving and reading
 // resources.
-func (r *Retriever) Retrieve() (map[string][]byte, error) {
+func (r *Retriever) Retrieve() (map[string][]byte, map[string]interface{}, error) {
 	var err error
 	r.logger.Info("Looking for spec-descriptors in 'spec'...")
 	for _, specDescriptor := range r.plan.CRDDescription.SpecDescriptors {
 		if err = r.read("spec", specDescriptor.Path, specDescriptor.XDescriptors); err != nil {
-			return nil, err
+			return nil, nil, err
```
Comment (otaviof) [inline pkg/controller/servicebindingrequest/retriever.go]: Same here, it should return a single unified context data.

File: pkg/controller/servicebindingrequest/secret.go [hunk 5]
```diff
@@ -20,12 +20,8 @@ type Secret struct {
 
 // customEnvParser parse informed data in order to interpolate with values provided by custom
 // environment component.
-func (s *Secret) customEnvParser(data map[string][]byte) (map[string][]byte, error) {
+func (s *Secret) customEnvParser(data map[string][]byte, cache map[string]interface{}) (map[string][]byte, error) {
```
Comment (isutton) [inline pkg/controller/servicebindingrequest/secret.go]: This method could be moved as is to `Retriever`and used by `Retrieve()`.

File: pkg/controller/servicebindingrequest/reconciler.go (hunk 3 above)
Comment (Avni-Sharma) [reply pkg/controller/servicebindingrequest/reconciler.go]: > Same idea here, make sure `retreievedData` contains it all.

But retrievedData maintains a key,value pair list whereas for cache we need to retain the whole structure (hierarchy) as mentioned here https://github.com/redhat-developer/service-binding-operator/issues/272#issuecomment-562735404

Comment (Avni-Sharma) [reply pkg/controller/servicebindingrequest/reconciler.go]: Having just the `secretData` would not maintain that structure

**Comment by sbose78** [comment]
Looks like the `customEnvVar` feature broke during our unbinding feature work https://github.com/redhat-developer/service-binding-operator/pull/256/files#diff-2c6d61012a9ded0b460e51c66cd47057R25

Since, fixing this the way suggested above is taking time, I would suggest taking in this PR to un-break the feature.

Here's what I would like to do :
1. Track the suggestion as a story.
2. Take this PR in.
3. @Avni-Sharma & @isutton work closely to understand what needs to be done. Take your time and improve the code.

At the same time, our tests need to improve so that we catch these issues.

What do you think?

**Comment by Avni-Sharma** [comment]
Tested locally
Cache hierarchy maintained. Output:
```
{"ignored key": {"spec":{"dbName":"db-demo","image":"docker.io/postgres","imageName":"postgres"},"status":{"db.host":{"db-demo":{}},"db.name":{"db-demo":{}},"db.password":{"db-demo":{}},"db.port":{"db-demo":{}},"db.user":{"db-demo":{}},"dbConfigMap":{"db.host":"172.30.89.99","db.name":"db-demo","db.password":"password","db.port":"5432","db.username":"postgres"},"dbConnectionIP":"172.30.89.99","dbConnectionPort":5432,"dbCredentials":"db-demo-postgresql","dbName":"db-demo","password":{},"user":{}}}}
```

Referred example https://github.com/redhat-developer/service-binding-operator/tree/master/examples/java_postgresql_customvar
Tested using devconsole UI

**Comment by Avni-Sharma** [comment]
```
status:
  applicationObjects:
  - service-binding-demo/java-app
  bindingStatus: Success
  secret: java-app-d-db-demo-postgresql-d
```

**Pull request merged**